app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -71,4 +97,132 @@ object MessageController {
         realm?.let(block) ?: RealmDatabase.mailboxContent().writeBlocking(block)
     }
     //endregion
+
+    //region New API routes
+    suspend fun fetchMessages(mailboxUuid: String, folderId: String) {
+
+        val folder = FolderController.getFolder(folderId) ?: return
+        val previousCursor = folder.cursor
+
+        val messagesUids = if (previousCursor == null) {
+            getMessagesUids(mailboxUuid, folderId)
+        } else {
+            getMessagesUidsDelta(mailboxUuid, folderId, previousCursor)
+        }
+
+        with(messagesUids) {
+
+            Log.e("API", "Threads - Added   : ${addedShortUids.count()}")
+            Log.e("API", "Threads - Deleted : ${deletedUids.count()}")
+            Log.e("API", "Threads - Updated : ${updatedMessages.count()}")
+
+            addMessages(addedShortUids, folder, mailboxUuid)
+
+            RealmDatabase.mailboxContent().writeBlocking {
+
+                deleteMessages(deletedUids, realm = this)
+                updateMessages(updatedMessages, mailboxUuid, folderId, realm = this)
+
+                if (cursor != null) FolderController.updateFolder(folderId, this) {
+                    it.lastUpdatedAt = Date().toRealmInstant()
+                    it.cursor = cursor
+                }
+            }
+        }
+
+        // TODO: Do we still need this with the new API routes?
+        // val isDraftFolder = FolderController.getFolder(folderId, this)?.role == FolderRole.DRAFT
+        // if (isDraftFolder) DraftController.cleanOrphans(threadsResult.threads, this)
+    }
+
+    private suspend fun addMessages(shortUids: List<String>, folder: Folder, mailboxUuid: String) {
+        if (shortUids.isNotEmpty()) {
+            val uids = getUniquesUidsInReverse(folder, shortUids)
+            val pageSize = ApiRepository.PER_PAGE
+            // val pageSize = 200 // TODO: Magic number
+            var pageStart = 0
+            while (pageStart < uids.count()) {
+                val pageEnd = min(pageStart + pageSize, uids.count())
+                val page = uids.subList(pageStart, pageEnd)
+                ApiRepository.getMessagesByUids(mailboxUuid, folder.id, page).data?.messages?.let { messages ->
+                    FolderController.updateFolder(folder.id) { folder ->
+                        folder.threads += messages.map { it.toThread(mailboxUuid) }.toRealmList()
+                    }
+                }
+                // TODO: Do we want a delay between each call, to not get blocked by the API?
+                delay(500L)
```

Comment:
🤔

Reply:
Seen with Workspace team.
The `delay` has been removed.